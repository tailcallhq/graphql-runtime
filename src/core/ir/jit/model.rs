use std::fmt::{Debug, Formatter};

use crate::core::ir::model::IR;

#[allow(unused)]
#[derive(Debug, Clone)]
pub struct Arg {
    pub id: ArgId,
    pub name: String,
    pub type_of: crate::core::blueprint::Type,
    pub value: Option<async_graphql_value::Value>,
    pub default_value: Option<async_graphql_value::ConstValue>,
}

#[derive(Clone)]
pub struct ArgId(usize);

impl Debug for ArgId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl ArgId {
    pub fn new(id: usize) -> Self {
        ArgId(id)
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct FieldId(usize);

impl Debug for FieldId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FieldId {
    pub fn new(id: usize) -> Self {
        FieldId(id)
    }
}

#[derive(Clone)]
pub struct Field<A: Clone> {
    pub id: FieldId,
    pub name: String,
    pub ir: Option<IR>,
    pub type_of: crate::core::blueprint::Type,
    pub args: Vec<Arg>,
    pub refs: Option<A>,
}

const EMPTY_VEC: &Vec<Field<Children>> = &Vec::new();
impl Field<Children> {
    #[allow(unused)]
    pub fn children(&self) -> &Vec<Field<Children>> {
        match &self.refs {
            Some(Children(children)) => children,
            _ => EMPTY_VEC,
        }
    }
}

impl Field<Parent> {
    fn parent(&self) -> Option<&FieldId> {
        self.refs.as_ref().map(|Parent(id)| id)
    }

    fn into_children(self, fields: &[Field<Parent>]) -> Field<Children> {
        let mut children = Vec::new();
        for field in fields.iter() {
            if let Some(id) = field.parent() {
                if *id == self.id {
                    children.push(field.to_owned().into_children(fields));
                }
            }
        }

        let refs = if children.is_empty() {
            None
        } else {
            Some(Children(children))
        };

        Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            args: self.args,
            refs,
        }
    }
}

impl<A: Debug + Clone> Debug for Field<A> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut debug_struct = f.debug_struct("Field");
        debug_struct.field("id", &self.id);
        debug_struct.field("name", &self.name);
        if self.ir.is_some() {
            debug_struct.field("ir", &"Some(..)");
        }
        debug_struct.field("type_of", &self.type_of);
        if !self.args.is_empty() {
            debug_struct.field("args", &self.args);
        }
        if self.refs.is_some() {
            debug_struct.field("refs", &self.refs);
        }
        debug_struct.finish()
    }
}

#[derive(Clone)]
pub struct Parent(FieldId);
#[allow(unused)]
impl Parent {
    pub fn new(id: FieldId) -> Self {
        Parent(id)
    }
}
impl Debug for Parent {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Parent({:?})", self.0)
    }
}

#[derive(Clone, Debug)]
pub struct Children(Vec<Field<Children>>);

#[derive(Clone, Debug)]
pub struct ExecutionPlan {
    parent: Vec<Field<Parent>>,
    children: Vec<Field<Children>>,
}

impl ExecutionPlan {
    pub fn new(fields: Vec<Field<Parent>>) -> Self {
        let field_children = fields
            .clone()
            .into_iter()
            .filter(|f| f.refs.is_none())
            .map(|f| f.into_children(&fields))
            .collect::<Vec<_>>();

        Self { parent: fields, children: field_children }
    }

    #[allow(unused)]
    pub fn as_children(&self) -> &[Field<Children>] {
        &self.children
    }

    #[allow(unused)]
    pub fn as_parent(&self) -> &[Field<Parent>] {
        &self.parent
    }

    #[allow(unused)]
    pub fn find_field(&self, id: FieldId) -> Option<&Field<Parent>> {
        self.parent.iter().find(|field| field.id == id)
    }
}
