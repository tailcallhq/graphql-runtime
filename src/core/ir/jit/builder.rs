use std::collections::HashMap;

use async_graphql::parser::types::{
    DocumentOperations, ExecutableDocument, OperationType, Selection, SelectionSet,
};

use super::model::*;
use crate::core::blueprint::{Blueprint, Index, QueryField};
use crate::core::counter::{Count, Counter};
use crate::core::merge_right::MergeRight;

#[allow(unused)]
pub struct Builder {
    pub index: Index,
    pub arg_id: Counter<usize>,
    pub field_id: Counter<usize>,
    pub document: ExecutableDocument,
}

#[allow(unused)]
impl Builder {
    pub fn new(blueprint: Blueprint, document: ExecutableDocument) -> Self {
        let index = blueprint.index();
        Self {
            document,
            index,
            arg_id: Counter::default(),
            field_id: Counter::default(),
        }
    }

    fn iter(
        &self,
        selection: &SelectionSet,
        type_of: &str,
        refs: Option<Parent>,
    ) -> Vec<Field<Parent>> {
        let mut fields = vec![];
        for selection in &selection.items {
            if let Selection::Field(gql_field) = &selection.node {
                let field_name = gql_field.node.name.node.as_str();
                let field_args = gql_field
                    .node
                    .arguments
                    .iter()
                    .map(|(k, v)| (k.node.as_str().to_string(), v.node.to_owned()))
                    .collect::<HashMap<_, _>>();

                if let Some(field_def) = self.index.get_field(type_of, field_name) {
                    let mut args = vec![];
                    for (arg_name, value) in field_args {
                        if let Some(arg) = field_def.get_arg(&arg_name) {
                            let type_of = arg.of_type.clone();
                            let id = ArgId::new(self.arg_id.next());
                            let name = arg_name.clone();
                            let default_value = arg
                                .default_value
                                .as_ref()
                                .and_then(|v| v.to_owned().try_into().ok());
                            args.push(Arg { id, name, type_of, value: Some(value), default_value });
                        }
                    }

                    let type_of = match field_def {
                        QueryField::Field((field_def, _)) => field_def.of_type.clone(),
                        QueryField::InputField(field_def) => field_def.of_type.clone(),
                    };

                    let id = FieldId::new(self.field_id.next());
                    let child_fields = self.iter(
                        &gql_field.node.selection_set.node,
                        type_of.name(),
                        Some(Parent::new(id.clone())),
                    );
                    let name = field_name.to_owned();
                    let ir = match field_def {
                        QueryField::Field((field_def, _)) => field_def.resolver.clone(),
                        _ => None,
                    };
                    fields.push(Field { id, name, ir, type_of, args, refs: refs.clone() });
                    fields = fields.merge_right(child_fields);
                }
            }
        }

        fields
    }

    fn get_type(&self, ty: OperationType) -> Option<&str> {
        match ty {
            OperationType::Query => Some(self.index.get_query()),
            OperationType::Mutation => self.index.get_mutation(),
            OperationType::Subscription => None,
        }
    }

    #[allow(unused)]
    pub fn build(&self) -> Result<ExecutionPlan, String> {
        let mut fields = Vec::new();

        for fragment in self.document.fragments.values() {
            let on_type = fragment.node.type_condition.node.on.node.as_str();
            fields.extend(self.iter(&fragment.node.selection_set.node, on_type, None));
        }

        match &self.document.operations {
            DocumentOperations::Single(single) => {
                let name = self.get_type(single.node.ty).ok_or(format!(
                    "Root Operation type not defined for {}",
                    single.node.ty
                ))?;
                fields.extend(self.iter(&single.node.selection_set.node, name, None));
            }
            DocumentOperations::Multiple(multiple) => {
                for single in multiple.values() {
                    let name = self.get_type(single.node.ty).ok_or(format!(
                        "Root Operation type not defined for {}",
                        single.node.ty
                    ))?;
                    fields.extend(self.iter(&single.node.selection_set.node, name, None));
                }
            }
        }

        Ok(ExecutionPlan::new(fields))
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::Config;
    use crate::core::ir::jit::builder::Builder;
    use crate::core::valid::Validator;

    const CONFIG: &str = include_str!("./fixtures/jsonplaceholder-mutation.graphql");

    fn plan(query: impl AsRef<str>) -> ExecutionPlan {
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let blueprint = Blueprint::try_from(&config.into()).unwrap();
        let document = async_graphql::parser::parse_query(query).unwrap();

        Builder::new(blueprint, document).build().unwrap()
    }

    #[tokio::test]
    async fn test_from_document() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[tokio::test]
    async fn test_size() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
        );

        assert_eq!(plan.size(), 4)
    }

    #[test]
    fn test_simple_query() {
        let plan = plan(
            r#"
            query {
                posts { user { id } }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_simple_mutation() {
        let plan = plan(
            r#"
            mutation {
              createUser(user: {
                id: 101,
                name: "Tailcall",
                email: "tailcall@tailcall.run",
                phone: "2345234234",
                username: "tailcall",
                website: "tailcall.run"
              }) {
                id
                name
                email
                phone
                website
                username
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_fragments() {
        let plan = plan(
            r#"
            fragment UserPII on User {
              name
              email
              phone
            }

            query {
              user(id:1) {
                ...UserPII
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_multiple_operations() {
        let plan = plan(
            r#"
            query {
              user(id:1) {
                id
                username
              }
              posts {
                id
                title
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_variables() {
        let plan = plan(
            r#"
            query user($id: Int!) {
              user(id: $id) {
                id
                name
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_unions() {
        let plan = plan(
            r#"
            query {
              getUserIdOrEmail(id:1) {
                ...on UserId {
                  id
                }
                ...on UserEmail {
                  email
                }
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }

    #[test]
    fn test_default_value() {
        let plan = plan(
            r#"
            mutation {
              createPost(post:{
                userId:123,
                title:"tailcall",
                body:"tailcall test"
              }) {
                id
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan);
    }
}
