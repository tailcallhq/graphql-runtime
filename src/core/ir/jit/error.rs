#[derive(Clone, Debug)]
pub enum Error {}

pub type Result<A> = std::result::Result<A, Error>;
