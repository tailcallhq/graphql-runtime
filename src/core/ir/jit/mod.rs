mod blueprint_index;
mod builder;
mod model;
