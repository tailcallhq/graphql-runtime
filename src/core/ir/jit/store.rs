use crate::core::ir::jit::model::FieldId;

#[allow(unused)]
pub struct Store<A> {
    map: Vec<Data<A>>,
}
#[allow(unused)]
#[derive(Clone)]
pub enum Data<A> {
    /// Represents that the value was computed only once for the associated
    /// field
    Single(A),
    /// Represents that the value was computed multiple times for the associated
    /// field. The order is guaranteed by the executor to be the same as the
    /// other of invocation and not the other of completion.
    Multiple(Vec<A>),
    /// Represents that the value is yet to be computed
    Pending,
}

impl<A> Store<A> {
    pub fn new(size: usize) -> Self {
        Store { map: (0..size).map(|_| Data::Pending).collect() }
    }

    pub fn set(&mut self, field_id: FieldId, data: Data<A>) {
        self.map.insert(field_id.as_usize(), data);
    }

    pub fn get(&self, field_id: &FieldId) -> Option<&Data<A>> {
        self.map.get(field_id.as_usize())
    }
}
