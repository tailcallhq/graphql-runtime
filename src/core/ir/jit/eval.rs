use derive_setters::Setters;
use value::ValueLike;

use super::*;
use crate::core::ir::model::*;
use crate::core::runtime::TargetRuntime;

/// NOTE: Cloning EvalContext should remain cheap for performance reasons.
#[allow(unused)]
#[derive(Clone, Setters)]
struct EvalContext<A> {
    #[setters(skip)]
    runtime: TargetRuntime,
    value: A,
    args: A,
    parent: A,
}

impl<A: ValueLike> EvalContext<A> {
    #[allow(unused)]
    pub fn new(runtime: TargetRuntime) -> Self {
        Self {
            runtime,
            value: A::default(),
            args: A::default(),
            parent: A::default(),
        }
    }

    #[allow(unused)]
    pub async fn eval(mut self, ir: IR) -> Result<A> {
        Box::pin(async move {
            match ir {
                IR::Context(ctx) => match ctx {
                    Context::Value => Ok(self.value.clone()),
                    Context::Path(path) => {
                        let a = path
                            .split_first()
                            .and_then(|(head, tail)| match head.as_str() {
                                "value" => self.value.path(tail),
                                "args" => self.args.path(tail),
                                "parent" => self.parent.path(tail),
                                _ => None,
                            });
                        Ok(a.unwrap_or(A::default()))
                    }
                    Context::PushArgs { expr, and_then } => {
                        let args = self.clone().eval(*expr).await?;
                        self.args(args).eval(*and_then).await
                    }
                    Context::PushValue { expr, and_then } => {
                        let value = self.clone().eval(*expr).await?;
                        self.value(value).eval(*and_then).await
                    }
                },
                IR::Dynamic(_) => todo!(),
                IR::IO(io) => match io {
                    IO::Http { req_template, group_by, dl_id, http_filter } => todo!(),
                    IO::GraphQL { req_template, field_name, batch, dl_id } => todo!(),
                    IO::Grpc { req_template, group_by, dl_id } => todo!(),
                    IO::Js { name } => todo!(),
                },
                IR::Cache(cache) => todo!(),
                IR::Path(_, _) => todo!(),
                IR::Protect(_) => todo!(),
                IR::Map(_) => todo!(),
            }
        })
        .await
    }
}
