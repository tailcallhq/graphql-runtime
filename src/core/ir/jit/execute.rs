use serde_json_borrow::OwnedValue;

use crate::core::ir::{EvaluationError, IR};

#[allow(unused)]
pub async fn execute_ir(
    ir: &IR,
    parent: Option<&OwnedValue>,
) -> Result<OwnedValue, EvaluationError> {
    todo!()
}
