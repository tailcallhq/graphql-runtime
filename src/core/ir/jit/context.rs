use std::sync::Mutex;

use futures_util::future;
use serde_json_borrow::OwnedValue;

use super::model::{ExecutionPlan, Field, FieldId, Parent};
use super::store::Store;
use crate::core::ir::{CallId, EvaluationError, IR};

#[allow(unused)]
pub struct ExecutionContext {
    plan: ExecutionPlan,
    store: Mutex<Store<CallId, OwnedValue>>,
}

#[allow(unused)]
impl ExecutionContext {
    pub async fn execute_ir(
        &self,
        ir: &IR,
        parent: Option<&OwnedValue>,
    ) -> Result<OwnedValue, EvaluationError> {
        todo!()
    }
    fn find_children(&self, id: FieldId) -> Vec<Field<Parent>> {
        todo!()
    }

    fn insert_field_value(&self, id: FieldId, value: OwnedValue) {
        todo!()
    }

    fn find_field(&self, id: FieldId) -> Option<&Field<Parent>> {
        self.plan.as_parent().iter().find(|field| field.id == id)
    }

    async fn execute_field(
        &self,
        id: FieldId,
        parent: Option<&OwnedValue>,
    ) -> Result<(), EvaluationError> {
        if let Some(field) = self.find_field(id.clone()) {
            if let Some(ir) = &field.ir {
                let value = self.execute_ir(ir, parent).await?;

                let children = self.find_children(id.clone());
                future::join_all(
                    children
                        .into_iter()
                        .map(|child| self.execute_field(child.id, Some(&value))),
                )
                .await
                .into_iter()
                .collect::<Result<Vec<_>, EvaluationError>>()?;

                self.insert_field_value(id, value);
            }
        }
        Ok(())
    }

    fn root(&self) -> Vec<&Field<Parent>> {
        self.plan
            .as_parent()
            .iter()
            .filter(|field| field.refs.is_none())
            .collect::<Vec<_>>()
    }

    pub async fn execute(self) -> Result<Store<CallId, OwnedValue>, EvaluationError> {
        future::join_all(
            self.root()
                .iter()
                .map(|field| self.execute_field(field.id.to_owned(), None)),
        )
        .await
        .into_iter()
        .collect::<Result<Vec<_>, EvaluationError>>()?;

        let store = self.store.into_inner().unwrap();

        Ok(store)
    }
}
