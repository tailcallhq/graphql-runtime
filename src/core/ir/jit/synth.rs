use serde_json_borrow::{ObjectAsVec, Value};

use crate::core::ir::jit::model::{Children, Field};
use crate::core::ir::jit::store::{Data, Store};

#[allow(unused)]
pub struct Synth<'a> {
    operations: Vec<Field<Children>>,
    store: Store<Value<'a>>,
}

#[allow(unused)]
impl<'a> Synth<'a> {
    pub fn new(operations: Vec<Field<Children>>, store: Store<Value<'a>>) -> Self {
        Self { operations, store }
    }
    pub fn synthesize(&self) -> Value {
        let mut data = ObjectAsVec::default();

        for child in self.operations.iter() {
            let val = self.iter(child, None, None);
            data.insert(child.name.as_str(), val);
        }

        let mut output = ObjectAsVec::default();
        output.insert("data", Value::Object(data));
        Value::Object(output)
    }

    /// checks if type_of is an array and value is an array
    fn is_array(type_of: &crate::core::blueprint::Type, value: &Value) -> bool {
        type_of.is_list() == value.is_array()
    }

    #[inline]
    fn iter<'b>(
        &'b self,
        node: &'b Field<Children>,
        parent: Option<&'b Value>,
        index: Option<usize>,
    ) -> Value {
        match parent {
            Some(parent) => {
                if !Self::is_array(&node.type_of, parent) {
                    return Value::Null;
                }
                self.iter_inner(node, Some(parent), index)
            }
            None => {
                // we perform this check to avoid unnecessary hashing

                match self.store.get(&node.id) {
                    Some(val) => {
                        match val {
                            // if index is given, then the data should be a list
                            // if index is not given, then the data should be a value
                            // must return Null in all other cases.
                            Data::Single(val) => {
                                if index.is_some() {
                                    return Value::Null;
                                }
                                self.iter(node, Some(val), None)
                            }
                            Data::Multiple(list) => {
                                if let Some(i) = index {
                                    match list.get(i) {
                                        Some(val) => self.iter(node, Some(val), None),
                                        None => Value::Null,
                                    }
                                } else {
                                    Value::Null
                                }
                            }
                            Data::Pending => {
                                // TODO: should bailout instead of returning Null
                                Value::Null
                            }
                        }
                    }
                    None => {
                        // IR exists, so there must be a value.
                        // if there is no value then we must return Null
                        Value::Null
                    }
                }
            }
        }
    }
    #[inline]
    fn iter_inner<'b>(
        &'b self,
        node: &'b Field<Children>,
        parent: Option<&'b Value>,
        index: Option<usize>,
    ) -> Value {
        match parent {
            Some(Value::Object(obj)) => {
                let mut ans = ObjectAsVec::default();
                let children = node.children();

                if children.is_empty() {
                    let val = obj.get(node.name.as_str());
                    // if it's a leaf node, then push the value
                    if let Some(val) = val {
                        ans.insert(node.name.as_str(), val.to_owned());
                    } else {
                        return Value::Null;
                    }
                } else {
                    for child in children {
                        let val = obj.get(child.name.as_str());
                        if let Some(val) = val {
                            ans.insert(
                                child.name.as_str(),
                                self.iter_inner(child, Some(val), index),
                            );
                        } else {
                            ans.insert(child.name.as_str(), self.iter(child, None, index));
                        }
                    }
                }
                Value::Object(ans)
            }
            Some(Value::Array(arr)) => {
                let mut ans = vec![];
                for (i, val) in arr.iter().enumerate() {
                    let val = self.iter_inner(node, Some(val), Some(i));
                    ans.push(val)
                }
                Value::Array(ans)
            }
            Some(val) => val.clone(), // cloning here would be cheaper than cloning whole value
            None => Value::Null,      // TODO: we can just pass parent value instead of an option
        }
    }
}

#[cfg(test)]
mod tests {

    use serde_json_borrow::Value;

    use crate::core::blueprint::Blueprint;
    use crate::core::config::{Config, ConfigModule};
    use crate::core::ir::common::JsonPlaceholder;
    use crate::core::ir::jit::builder::Builder;
    use crate::core::ir::jit::model::FieldId;
    use crate::core::ir::jit::store::{Data, Store};
    use crate::core::ir::jit::synth::Synth;
    use crate::core::valid::Validator;

    const POSTS: &str = r#"
        [
                {
                    "id": 1,
                    "userId": 1,
                    "title": "Some Title"
                },
                {
                    "id": 2,
                    "userId": 1,
                    "title": "Not Some Title"
                }
        ]
    "#;

    const USER1: &str = r#"
        {
                "id": 1,
                "name": "foo"
        }
    "#;

    const USER2: &str = r#"
        {
                "id": 2,
                "name": "bar"
        }
    "#;
    const USERS: &str = r#"
        [
          {
            "id": 1,
            "name": "Leanne Graham"
          },
          {
            "id": 2,
            "name": "Ervin Howell"
          }
        ]
    "#;

    enum TestData {
        Posts,
        UsersData,
        Users,
        User1,
    }

    impl TestData {
        fn into_value(self) -> Data<Value<'static>> {
            match self {
                Self::Posts => Data::Single(serde_json::from_str(POSTS).unwrap()),
                Self::User1 => Data::Single(serde_json::from_str(USER1).unwrap()),
                TestData::UsersData => Data::Multiple(vec![
                    serde_json::from_str(USER1).unwrap(),
                    serde_json::from_str(USER2).unwrap(),
                ]),
                TestData::Users => Data::Single(serde_json::from_str(USERS).unwrap()),
            }
        }
    }

    const CONFIG: &str = include_str!("./fixtures/jsonplaceholder-mutation.graphql");

    fn init(query: &str, store: Vec<(FieldId, Data<Value<'static>>)>) -> String {
        let doc = async_graphql::parser::parse_query(query).unwrap();
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let config = ConfigModule::from(config);

        let builder = Builder::new(Blueprint::try_from(&config).unwrap(), doc);
        let plan = builder.build().unwrap();

        let store = store
            .into_iter()
            .fold(Store::new(plan.size()), |mut store, (id, data)| {
                store.set(id, data);
                store
            });

        let synth = Synth::new(plan.into_children(), store);
        let val = synth.synthesize();

        serde_json::to_string_pretty(&val).unwrap()
    }

    #[test]
    fn test_posts() {
        let store = vec![(FieldId::new(0), TestData::Posts.into_value())];

        let val = init(
            r#"
            query {
                posts { id }
            }
        "#,
            store,
        );
        insta::assert_snapshot!(val);
    }

    #[test]
    fn test_user() {
        let store = vec![(FieldId::new(0), TestData::User1.into_value())];

        let val = init(
            r#"
            query {
                user(id: 1) { id }
            }
        "#,
            store,
        );
        insta::assert_snapshot!(val);
    }

    #[test]
    fn test_nested() {
        let store = vec![
            (FieldId::new(0), TestData::Posts.into_value()),
            (FieldId::new(3), TestData::UsersData.into_value()),
        ];

        let val = init(
            r#"
            query {
                posts { id title user { id name } }
            }
        "#,
            store,
        );
        insta::assert_snapshot!(val);
    }

    #[test]
    fn test_multiple_nested() {
        let store = vec![
            (FieldId::new(0), TestData::Posts.into_value()),
            (FieldId::new(3), TestData::UsersData.into_value()),
            (FieldId::new(6), TestData::Users.into_value()),
        ];

        let val = init(
            r#"
            query {
                posts { id title user { id name } }
                users { id name }
            }
        "#,
            store,
        );
        insta::assert_snapshot!(val)
    }

    #[test]
    fn test_json_placeholder() {
        // FIXME: doesn't work when userId is queried
        let synth = JsonPlaceholder::init("{ posts { id title userId user { id name } } }");
        let val = synth.synthesize();
        insta::assert_snapshot!(serde_json::to_string_pretty(&val).unwrap())
    }
}
