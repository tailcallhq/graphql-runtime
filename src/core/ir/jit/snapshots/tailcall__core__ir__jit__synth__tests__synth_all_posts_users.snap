---
source: src/core/ir/jit/synth.rs
expression: actual
---
{
  "posts": [
    {
      "id": 1,
      "title": "My title",
      "user": {
        "name": "Jane Doe"
      }
    },
    {
      "id": 2,
      "title": "Also My Title",
      "user": {
        "name": "Jane Doe"
      }
    }
  ]
}
